modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java

```diff
@@ -733,41 +745,68 @@ public static <RowT> RowT validateCharactersOverflowAndTrimIfPossible(
 
         for (int i = 0; i < colCount; ++i) {
             RelDataType colType = rowType.getFieldList().get(i).getType();
+            SqlTypeName typeName = colType.getSqlTypeName();
             Object data = rowHandler.get(i, row);
 
-            // Skip null values and non-character types.
-            if (!CHAR_TYPES.contains(colType.getSqlTypeName()) || data == null) {
+            if (data == null || (!BINARY_TYPES.contains(typeName) && !CHAR_TYPES.contains(typeName))) {
                 if (rowBldr != null) {
                     rowBldr.addField(data);
                 }
                 continue;
             }
-            // Otherwise validate and trim if needed.
 
-            assert data instanceof String;
+            int colPrecision = colType.getPrecision();
 
-            String str = (String) data;
+            // Validate and trim if needed.
 
-            int colPrecision = colType.getPrecision();
+            if (BINARY_TYPES.contains(typeName)) {
+                assert data instanceof ByteString;
+                if (typeName == SqlTypeName.VARBINARY && colType.getPrecision() == RelDataType.PRECISION_NOT_SPECIFIED) {
+                    continue;
+                }
+
+                ByteString byteString = (ByteString) data;
+
+                if (byteString.length() > colPrecision) {
+                    for (int pos = byteString.length(); pos > colPrecision; --pos) {
+                        if (byteString.byteAt(pos - 1) != 0) {
+                            throw new SqlException(STMT_VALIDATION_ERR, "Value too long for type: " + colType);
+                        }
+                    }
 
-            assert colPrecision != RelDataType.PRECISION_NOT_SPECIFIED;
```

Comment:
this is correct assertion. At this point we must not get colType.getPrecision() == RelDataType.PRECISION_NOT_SPECIFIED

Reply:
add this check for both